pandas/core/groupby/groupby.py

```diff
@@ -3290,6 +3304,102 @@ def max(
                 npfunc=np.max,
             )
 
+    def _idxmin_idxmax(
+        self,
+        how: Literal["idxmax", "idxmin"],
+        numeric_only: bool = False,
+        skipna: bool = True,
+        ignore_unobserved: bool = False,
+    ) -> NDFrameT:
+        """Compute idxmax/idxmin.
+
+        Parameters
+        ----------
+        how: {"idxmin", "idxmax"}
+            Whether to compute idxmin or idxmax.
+        numeric_only : bool, default False
+            Include only float, int, boolean columns.
+        skipna : bool, default True
+            Exclude NA/null values. If an entire row/column is NA, the result
+            will be NA.
+        ignore_unobserved : bool, default False
+            When True and an unobserved group is encountered, do not raise. This used
+            for transform where unobserved groups do not play an impact on the result.
+
+        Returns
+        -------
+        Series or DataFrame
+            idxmax or idxmin for the groupby operation.
+        """
+        # seen is passed to the Cython code and mutated
+        seen = np.zeros(self.ngroups, dtype="bool")
+
+        def post_process(res):
+            has_na_value = (res._values == -1).any(axis=None)
+            has_unobserved = not seen.all()
+            raise_err: bool | np.bool_ = not ignore_unobserved and has_unobserved
+            if (
+                not self.observed
+                and not ignore_unobserved
+                and not raise_err
+                and (len(self.grouper.groupings) > 1 or res.empty)
+            ):
+                # When there are multiple groupings or the result is empty, the
+                # categories are not included in the upfront computation so we
+                # compare the final result length with the current length
+                result_len = np.prod(
+                    [len(ping.group_index) for ping in self.grouper.groupings]
+                )
+                raise_err = len(res) < result_len
```

Comment:
This should raise on `!=` right?

Reply:
Yea, I prefer to add an assert for <= here rather than test for != so we don't raise a confusing error. Will add.

Reply:
Sorry I wasn't super clear but I was wondering about the possiblity that `len(res) > result_len`

Reply:
Oh - I thought you were clear :smile:. We should always have `len(res) <= result_len`  without exception. I've added an assert to this effect. Does that resolve this?

Reply:
Oh OK I see what you did. Not to be too pedantic but can we change that to `assert not len(res) > result_len`? Just from a readability perspective it took me a few reads to understand the difference in intent between this and the line following

Reply:
This does indeed seem too pedantic to me. There is nothing wrong with the comparability of the two lines as far as I can tell.